### PR TITLE
kernel: queue: add several test cases of error scenario

### PR DESCRIPTION
**Pull request by Zhaoningx**
Add some testcases to test some failure scenario
to enhance the coverage of queue's source code.
And add the fatal error function to handler the
fatal error by ourself.

Signed-off-by: Ningx Zhao <ningx.zhao@intel.com>

### CODE REVIEW

**Review by LeiW000 — CHANGES_REQUESTED** [review]
I think we still need to add some details with "@details" doxygen tag for the test cases.

File: tests/kernel/queue/src/test_queue_contexts.c [hunk 1]
```diff
@@ -28,13 +28,19 @@ static struct k_thread tdata;
...
 {
 	if (reason != K_ERR_KERNEL_OOPS) {
 		printk("Wrong error type\n");
 		k_fatal_halt(reason);
 	}
+	printk("Caught system error -- reason %d %d\n", reason, valid_fault);
+	if (valid_fault) {
+		printk("Fatal error expected as part of test case.\n");
+		valid_fault = false; /* reset back to normal */
+	}
```
Comment (enjiamai) [inline tests/kernel/queue/src/test_queue_contexts.c:43]: What will happen if the "reason==K_ERR_KERNEL_OOPS" and "valid_fault==false"?

**Review by enjiamai — CHANGES_REQUESTED** [review]
I have two suggestions: 
1. The title seems not that good. Maybe something like: "add several test cases of error scenario" would be better.
2. You should descript more information in the @details field, at least what this test case want to validate and how you do this, etc.

File: tests/kernel/queue/src/test_queue_contexts.c (hunk 1 above)
Comment (Zhaoningx) [reply tests/kernel/queue/src/test_queue_contexts.c:43]: It seems that it will do nothing in this function.

**Comment by Zhaoningx** [comment]
Hi @LeiW000  @enjiamai , I have updated and re-pushed it. Could you review it again, thanks.

**Review by LeiW000 — APPROVED** [review]

**Review by enjiamai — APPROVED** [review]

File: tests/kernel/queue/src/test_queue_contexts.c [hunk 2]
```diff
@@ -28,10 +28,15 @@ static struct k_thread tdata;
 static K_THREAD_STACK_DEFINE(tstack1, STACK_SIZE);
 static struct k_thread tdata1;
 static struct k_sem end_sema;
+extern ZTEST_BMEM volatile bool valid_fault;
 
 void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *pEsf)
 {
-	if (reason != K_ERR_KERNEL_OOPS) {
+	printk("Caught system error -- reason %d %d\n", reason, valid_fault);
```
Comment (andrewboie) [inline tests/kernel/queue/src/test_queue_contexts.c:35]: same question as in other PR....I remember reviewing a PR which adds a generic fatal error handling framework in tests, can we use that?

**Review by enjiamai — COMMENTED** [review]
1. The common fatal and assert handler has been merged, could you please leverage them instead?
2. And let's remove error test case which trigger assertion which passing NULL directly, but keep others which will trigger Z_OOPS() from userspace.  
please refer to https://github.com/zephyrproject-rtos/zephyr/pull/30150
Thanks!

**Review by enjiamai — CHANGES_REQUESTED** [review]
1. The common fatal and assert handler has been merged, could you please leverage them instead?
2. And let's remove error test case which trigger assertion which passing NULL directly, but keep others which will trigger Z_OOPS() from userspace.  
please refer to https://github.com/zephyrproject-rtos/zephyr/pull/30150
Thanks!

File: tests/kernel/queue/src/test_queue_contexts.c (hunk 2 above)
Comment (Zhaoningx) [reply tests/kernel/queue/src/test_queue_contexts.c]: HI @andrewboie , after a generic fatal error handling framework got merged, I have modified my code and re-push it, so could you take look at this PR again, thanks.

**Comment by Zhaoningx** [comment]
Hi @enjiamai , I have modified and could you have a take look at this PR?

**Review by enjiamai — APPROVED** [review]

**Comment by Zhaoningx** [comment]
Hi @andyross , the PR has got two approvals, so could you take a look at this, thanks.

**Review by nashif — APPROVED** [review]

**Pull request merged**